Make PlatformMediaSessionClient a CheckedPtr

jernoble opened the pull request:
#### e5e34e38459892dbee6c0941c00d15500eed6ce0
<pre>
Make PlatformMediaSessionClient a CheckedPtr
<a href="https://rdar.apple.com/149224332">rdar://149224332</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=291585">https://bugs.webkit.org/show_bug.cgi?id=291585</a>

Reviewed by Andy Estes.

Make PlatformMediaSessionClient a CheckedPtr and add an explicit invalidate() client
call to PlatformMediaSessionInterface.

Add a new empty PlatformMediaSessionClient subclass for use when the main client is
invalidated that uses a (also new) empty Logger. This keeps clients from having to
test that the Logger is non-null through LOG_IF_POSSIBLE macros.

SaferCpp Bot Fixes:
- Now that HTMLMediaElement is _explicitly_ capable of being a CheckedPtr (previously
  it was only implicitly CheckedPtr via Node), fix a number of SaferCpp warnings about
  raw or WeakPtr use of HTMLMediaElement.

* Source/WTF/wtf/Logger.cpp:
(WTF::emptyLogger):
* Source/WTF/wtf/Logger.h:
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::~AudioContext):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::~HTMLMediaElement):
* Source/WebCore/html/HTMLMediaElement.h:
(WebCore::HTMLMediaElement::checkedPtrCount const):
(WebCore::HTMLMediaElement::checkedPtrCountWithoutThreadCheck const):
(WebCore::HTMLMediaElement::incrementCheckedPtrCount const):
(WebCore::HTMLMediaElement::decrementCheckedPtrCount const):
* Source/WebCore/platform/audio/PlatformMediaSessionInterface.cpp: Added.
(WebCore::emptyPlatformMediaSessionClient):
* Source/WebCore/platform/audio/PlatformMediaSessionInterface.h:
(WebCore::PlatformMediaSessionInterface::invalidateClient):

Canonical link: <a href="https://commits.webkit.org/293979@main">https://commits.webkit.org/293979@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7b81e04164a09ee88496f661f8387e5ae2901638

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100443 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20095 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10394 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105580 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51031 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102484 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20402 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28569 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76483 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33535 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103450 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15635 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90732 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56839 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15451 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8735 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50406 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/93107 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85368 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8814 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107934 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/99051 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27561 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20227 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/85443 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27924 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86932 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84981 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21619 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29658 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7391 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21526 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27496 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32746 "Found 32 new failures in platform/audio/PlatformMediaSessionInterface.h, platform/audio/PlatformMediaSession.cpp and found 2 fixed files: mac/DOM/DOMHTMLVideoElement.mm, mac/DOM/DOMHTMLMediaElement.mm") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/122677 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27307 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34220 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30625 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28865 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WTF/wtf/Logger.cpp:
```diff
@@ -68,4 +68,16 @@ Vector<std::reference_wrapper<Logger::MessageHandlerObserver>>& Logger::messageH
 
+const Logger& emptyLogger()
+{
+    static NeverDestroyed<Ref<Logger>> emptyLogger = [&] {
```
Nit: Since `Logger` is already heap-allocated there's no benefit to using `NeverDestroyed`. We could just leak it instead.

jernoble replied:
@aestes We discussed offline how the other benefit of using NeverDestroyed is that we can use its location as the "owner" of the `Logger`.

nt1m commented on Source/WebCore/Modules/mediacontrols/MediaControlsHost.cpp:
```diff
@@ -145,10 +145,11 @@ const AtomString& MediaControlsHost::mediaControlsContainerClassName() const
 {
-    if (!m_mediaElement)
+    RefPtr mediaElement = protectedElement();
+    if (!mediaElement)
         return { };
 
-    Page* page = m_mediaElement->document().page();
+    Page* page = mediaElement->document().page();
```
The current code is initializing some unnecessary `RefPtr`. Same applies to a lot of this file

Null checks don't really need to be protected. So this can just be written as:
```
if (!m_mediaElement)
    return { };

Page* page = protectedElement()->document().page();
```

(This isn't really relevant with my suggestion, but `protectedElement()` should really be reserved to when there's a single usage of the method, e.g. `protectedElement()->foo()`, if you're storing it inside a RefPtr variable, you can use the non protected function, since the RefPtr variable protects it already):

```
RefPtr mediaElement = m_mediaElement;
if (!mediaElement)
    return { };

Page* page = mediaElement->document().page();
```

jernoble replied:
Initializing a RefPtr with a null pointer is nearly free; this is a micro-optimization that relies upon a subtle understanding of the underlying code, and an assumption that there can be nothing between the null or weak check which can be violated by future authors.

I added and used `protectedElement()` so that I could easily tell by code inspection that no unsafe uses of the raw `m_mediaElement` remained. Again, this makes this code robust against future authors, by developing a pattern where the raw or weak pointer is not accessed directly.

nt1m commented on Source/WebCore/Modules/mediacontrols/MediaControlsHost.cpp:
```diff
@@ -217,14 +221,16 @@ AtomString MediaControlsHost::captionDisplayMode() const
 {
-    if (m_mediaElement)
-        m_mediaElement->setSelectedTextTrack(track);
+    RefPtr mediaElement = protectedElement();
+    if (mediaElement)
+        mediaElement->setSelectedTextTrack(track);
```
You could write this in two ways:
```
    if (m_mediaElement)
        protectedElement()->setSelectedTextTrack(track);
```
or
```
    if (RefPtr mediaElement = m_mediaElement)
        mediaElement->setSelectedTextTrack(track);
```

jernoble replied:
I prefer using the `protectedElement()` method as, per above, it removes the pattern where the raw pointer is accessed.

jernoble replied:
I also checked with godbolt and verified that the proposed two options generate the exact same machine code; given there does not seem to be even a micro-optimization here, I'd prefer to use the `if (Ref = protected)` pattern for the reasons mentioned above.

nt1m commented on Source/WebCore/Modules/mediacontrols/MediaControlsHost.cpp:
```diff
@@ -217,14 +221,16 @@ AtomString MediaControlsHost::captionDisplayMode() const
 {
-    if (!m_textTrackContainer && m_mediaElement)
-        m_textTrackContainer = MediaControlTextTrackContainerElement::create(m_mediaElement->document(), *m_mediaElement);
+    RefPtr mediaElement = protectedElement();
+    if (!m_textTrackContainer && mediaElement)
+        m_textTrackContainer = MediaControlTextTrackContainerElement::create(mediaElement->document(), *mediaElement);
```
```suggestion
    if (!m_textTrackContainer && m_mediaElement)
        m_textTrackContainer = MediaControlTextTrackContainerElement::create(mediaElement->document(), *protectedElement());
```

jernoble replied:
You've left a `mediaElement` in the `create(...)` method. To follow this pattern would require two uses of `protectedElement()`, requiring separate RefPtrs to be created.

nt1m commented on Source/WebCore/Modules/mediacontrols/MediaControlsHost.cpp:
```diff
@@ -274,46 +280,54 @@ void MediaControlsHost::updateCaptionDisplaySizes(ForceUpdate force)
 bool MediaControlsHost::allowsInlineMediaPlayback() const
 {
-    return m_mediaElement && !m_mediaElement->mediaSession().requiresFullscreenForVideoPlayback();
+    RefPtr mediaElement = protectedElement();
+    return mediaElement && !mediaElement->mediaSession().requiresFullscreenForVideoPlayback();
```
Here and below:

```suggestion
    return m_mediaElement && !protectedElement()->mediaSession().requiresFullscreenForVideoPlayback();
```

jernoble replied:
Again, this is a micro-optimization that isn't justified.

jernoble committed:
Make PlatformMediaSessionClient a CheckedPtr
rdar://149224332
https://bugs.webkit.org/show_bug.cgi?id=291585

Reviewed by Andy Estes.

Make PlatformMediaSessionClient a CheckedPtr and add an explicit invalidate() client
call to PlatformMediaSessionInterface.

Add a new empty PlatformMediaSessionClient subclass for use when the main client is
invalidated that uses a (also new) empty Logger. This keeps clients from having to
test that the Logger is non-null through LOG_IF_POSSIBLE macros.

SaferCpp Bot Fixes:
- Now that HTMLMediaElement is _explicitly_ capable of being a CheckedPtr (previously
  it was only implicitly CheckedPtr via Node), fix a number of SaferCpp warnings about
  raw or WeakPtr use of HTMLMediaElement.

* Source/WTF/wtf/Logger.cpp:
(WTF::emptyLogger):
* Source/WTF/wtf/Logger.h:
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::~AudioContext):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::~HTMLMediaElement):
* Source/WebCore/html/HTMLMediaElement.h:
(WebCore::HTMLMediaElement::checkedPtrCount const):
(WebCore::HTMLMediaElement::checkedPtrCountWithoutThreadCheck const):
(WebCore::HTMLMediaElement::incrementCheckedPtrCount const):
(WebCore::HTMLMediaElement::decrementCheckedPtrCount const):
* Source/WebCore/platform/audio/PlatformMediaSessionInterface.cpp: Added.
(WebCore::emptyPlatformMediaSessionClient):
* Source/WebCore/platform/audio/PlatformMediaSessionInterface.h:
(WebCore::PlatformMediaSessionInterface::invalidateClient):

Canonical link: https://commits.webkit.org/293979@main
